internal/c/parts/audio/audio.cpp

```diff
@@ -703,61 +622,47 @@ static ma_uint8 *GenerateWaveform(double frequency, double length, double volume
                 value = -2.0 - value;
             }
         }
-    } // i
+    }
 
+    // Update reduced size if waveend is non-zero
     if (waveend)
-        *soundwave_bytes = waveend * SAMPLE_FRAME_SIZE(ma_int16, 2);
+        *soundwave_bytes = waveend * SAMPLE_FRAME_SIZE(float, 1);
 
     return data;
 }
 
-/// <summary>
-/// Returns the of a sound buffer in bytes.
-/// </summary>
-/// <param name="length">Length in seconds</param>
-/// <returns>Length in bytes</returns>
-static ma_int32 WaveformBufferSize(double length) {
-    static ma_int32 samples;
-
-    samples = (ma_int32)(length * audioEngine.sampleRate);
+/// @brief Returns the of a sound buffer in bytes
+/// @param length Length in seconds
+/// @return Length in bytes
+static int WaveformBufferSize(double length) {
+    auto samples = (int)(length * audioEngine.sampleRate);
     if (!samples)
         samples = 1;
 
-    return samples * SAMPLE_FRAME_SIZE(ma_int16, 2);
+    return samples * SAMPLE_FRAME_SIZE(float, 1);
 }
 
-/// <summary>
-/// This sends a buffer to a raw queue for playback.
-/// Buffer required in 16-bit stereo at native frequency.
-/// The buffer is freed.
-/// </summary>
-/// <param name="data">Sound buffer</param>
-/// <param name="bytes">Length of buffer in bytes</param>
-/// <param name="block">So we have to wait until playback completes</param>
-/// <param name="sndRawQueue">A pointer to a raw queue object</param>
-static void SendWaveformToQueue(ma_uint8 *data, ma_int32 bytes, bool block) {
-    static ma_int32 i;
-    static ma_int64 time_ms;
-
+/// @brief This sends a FP32 mono buffer to a raw stream for playback and then frees the buffer
+/// @param data FP32 mono sound buffer
+/// @param bytes Length of buffer in bytes
+/// @param block If this is true the function is wait until the wavefrom has finished playing
+static void SendWaveformToQueue(float *data, int bytes, bool block) {
     if (!data)
         return;
 
+    auto samples = bytes / SAMPLE_FRAME_SIZE(float, 1);
+
     // Move data into sndraw handle
-    for (i = 0; i < bytes; i += SAMPLE_FRAME_SIZE(ma_int16, 2)) {
-        audioEngine.soundHandles[audioEngine.sndInternal]->rawQueue->PushSampleFrame((float)((ma_int16 *)(data + i))[0] / 32768.0f,
-                                                                                     (float)((ma_int16 *)(data + i))[1] / 32768.0f);
+    for (auto i = 0; i < samples; i++) {
```

Comment:
If this turns out to be a bit slow we could add a `PushSampleFrames()` function which takes the full buffer and length, then we would only need to lock the mutex once for the whole copy, rather than once per sample. Since this is on the producer side performance is less of a concern though.

Reply:
Totally agree. I'll add that just in case. However, we cannot use `PushSampleFrames()` for `sub__sndraw` since it only pushes one frame at a time.